# 01 - Positional Arguments

In this context, `a` and `b` are called **parameters** of `my_func`. `a` and `b` are also variables local to `my_func`.

In [96]:
def my_func(a, b):
    # code here
    pass

`x` and `y` are called the **arguments** of my_func. Also importantly, `x` and `y` are passed by **reference (memory address)** so the variables `a` and `b` are pointing to the same memory address as `x` and `y`.

In [97]:
x = 10
y = 'a'

my_func(x, y)

**Python rules:**

- **If a positional parameter is defined with a default value, every positional parameter after it must also be given a default value, otherwise, you'll get an error.** 
- **Once you use a named argument, all arguments thereafter must be named too.**

In [98]:
def my_func(a, b=100, c) # rule 1

my_func(1, b=2, 3) # rule 2

SyntaxError: non-default argument follows default argument (2926666543.py, line 1)

# 02 - Unpacking Iterables

A note about tuples: A tuple is not defined by the `()` but instead by `,`. Brackets are used for clarity, except when creating an empty tuple in which case you can do `()`. For more than 0, we can write `1,` or `(1,)` but **not** `(1)`.

Unpacking is the act of splitting packed values (iterables, which includes strings) into individual variables. They are unpacked based on the relative positions of each element.

In [99]:
a, b, c = [1, 2, 3] # The LHS is a tuple.

In [100]:
a, b, c = 'XYZ'

This is very useful for swapping values of two variables because we don't need a temporary variable. To swap `a` and `b`:

In [101]:
tmp = a
a = b      # bad method
b = tmp

a, b = b, a # good method

**Unpacking sets and dictionaries**
Iterating using `for <> in` iterates through the keys.

In [1]:
d = {'key1': 1, 'key2':2, 'key3':3}

for i in d:
    print(i)

key1
key2
key3


In earlier versions of python, unpacking a dictionary did not preserve order but now it does in their insertion order.

In [103]:
a, b, c = d
print(a, b, c)

key1 key2 key3


But not for sets!

In [104]:
s = {'p', 'y', 't', 'h', 'o', 'n'}
a, b, c, d, e, f  =s
print(a, b, c, d, e, f)

h p n y o t


# 03 - Extended Unpacking

**Single star operator `*`**

If we don't want to unpack every single item in an iterable, we can assign the remaining variables to a list using `*` ; even if we unpack a tuple, the remainder is unpacked into a list. \
The two ways we can do it are:

In [105]:
list1 = [1, 2, 3, 4, 5, 6]
list2 = [1, 2, 3, 4, 5, 6]

a, b = list1[0], list1[1:]
c, *d = list2

print(a, b)
print(c, d)

1 [2, 3, 4, 5, 6]
1 [2, 3, 4, 5, 6]


The second way is good because it works with **any iterable**, not just sequence types

In [106]:
a, *b = 'XYZ'
print(a, b)

X ['Y', 'Z']


This also allows a shorthand for unpacking a string into a list without using the `list()` constructor:

In [107]:
*a, = 'WXYZ'
print(a)

['W', 'X', 'Y', 'Z']


We can even have `*` in the middle of unpacking.

In [108]:
a, b, *c, d = [1, 2, 3, 4, 5]
print(a, b, c, d)

1 2 [3, 4] 5


We can also unpack on the right hand side. If we want to empty the contents of two lists into a larger list without nesting lists, we can write:

In [109]:
list1 = [1, 2, 3]
list2 = [4, 5, 6]
l = [*list1, *list2]
print(l)

[1, 2, 3, 4, 5, 6]


We can unpack dictionaries with `*` on the LHS and RHS, but remember, when we iterate through a dictionary, we iterate through the keys only, so those are the variables that are unpacked.

In [110]:
d1 = {'p': 1, 'y': 2}
d2 = {'t': 3, 'h':4, 'o':5, 'n':6}

l = [*d1, *d2]
print(l)

['p', 'y', 't', 'h', 'o', 'n']


The `*` operator is very useful for converting a list or tuple into comma-separated arguments.

In [2]:
my_list = ['flower', 'flood', 'flee']

for i in zip(*my_list): # identical to zip('flower', 'flood', 'flee')
    print(i)

('f', 'f', 'f')
('l', 'l', 'l')
('o', 'o', 'e')
('w', 'o', 'e')


Remember, strings are iterables so zipping allows us to take the first element of each argument in the zip and pack it in a tuple. Then the second element etc. 

**Double star operator `**`**

If we want to unpack the keys and values, we can use the `**` operator. This is useful for merging dictionaries. If a key is repeated, then the last unpacking will take precedence (see `'h'` below):

In [111]:
d1 = {'p': 1, 'y': 2}
d2 = {'t': 3, 'h':4}
d3 = {'h': 5, 'o':6, 'n':7}

d = {**d1, **d2, **d3}
print(d)


{'p': 1, 'y': 2, 't': 3, 'h': 5, 'o': 6, 'n': 7}


So, if we had a dictionary with 20 keys and we wanted to update 5 of them, we could just create a dictionary with the 5 and merge it with the first dictionary.

In [112]:
to_be_merged = {'a':5, 'c':7}
big_dictionary = {'a':1, 'b':2, 'c':3, **to_be_merged}
print(big_dictionary)

{'a': 5, 'b': 2, 'c': 7}


**Nested unpacking**

if we had `l = [1, 2, [3, 4]]`, how would we unpack everything into individual variables? Using the tuple way:

In [113]:
l = [1, 2, [3, 4]]

In [114]:
a, b, (c, d) = l  # To make it easier to see, you can replace [3,4] with x and (c, d) with x
print(a, b, c, d)

1 2 3 4


Although we said we can't use `*` twice on LHS, we can if the `*` are at different nesting levels:

In [115]:
a, *b, (c, *d) = [1, 2, 3, 'python']
print(a, b, c, d)
# Note (c, *d) can be understood as: c, *d = 'python'.


1 [2, 3] p ['y', 't', 'h', 'o', 'n']


# 04 - Star-Args 

A similar thing can be done with positional arguments that are passed to a function.\
**One key difference is that the `*` unpacks to a tuple and *not* a list**

In [116]:
def func1(a, b, *c): # It is convention to write *args instead of *c.
    return c

func1(20, 10, 1, 2, 3, 'a', 'b')

(1, 2, 3, 'a', 'b')

One inconsistency with the above unpacking is that we can't have positional arguments after `*args`:

In [117]:
def func(a, b, *args, d):
    print (args)

func(1, 2, 3, 4, 5, d) # but will work if d=6 used instead, because it comes after *, therefore, it is mandatory.

TypeError: func() missing 1 required keyword-only argument: 'd'

Unless d is called by a keyword argument `d=6`.

**Unpacking arguments** \
If we have a list `l = [10, 20, 30]` and we want to pass each value of that list to three separate parameters, we can use unpacking.

In [118]:
l = [10, 20, 30]

def func1(a, b, c):
    print(a, b, c)
    
func1(*l)

10 20 30


# 05 - Keyword Arguments

When it came to passing arguments, it was entirely up to the user whether they wanted to use positional or keyword arguments. But what if wanted **mandatory keyword arguments**? \
To do so, we need to make sure that we've exhausted all the positional parameters first.

In [119]:
def func(a, b, *args, d):
    pass

Here, the first two arguments passed will be set to `a` and `b`, respectively. The remainder of the args will be stored in the `args` tuple, and finally, d is a keyword argument that **must** be specified.

We can actually force no positional arguments using just `*`: The * indicates to python that this is the "end" of positional arguments; everything after will be keyword arguments. Since there are no args before that, we effectively have no positional arguments.

If we have 2 parameters followed by * which is followed by `N` keyword parameters, then we **must** pass exactly 2 positional arguments and exactly `N` keyword arguments.

In [120]:
def func(*, kw1, kw2): # * indicates the "end" of positional arguments
    print(kw1, kw2)

#func(5, 10, 15) # Fails: TypeError: func() takes 0 positional arguments but 3 were given
#func(kw1=5)     # Fails: TypeError: func() missing 1 required keyword-only argument: 'kw2'
func(kw1=5, kw2=10)

5 10


In [121]:
def func(a, *, kw1, kw2): # * indicates the "end" of positional arguments
    print(kw1, kw2)

# func(kw1=5, kw2=10) # Fails: TypeError: func() missing 1 required positional argument: 'a'
func(5, kw1=5, kw2=10)

5 10


Don't think that using keyword arguments will solve any positional related issues. Sometimes you must use a positional argument than a keyword. In the below, we have passed a keyword argument `a=2`, but then followed it with positional arguments - this is **not allowed**.

To fix it, replace `a=2` with `2`.

In [122]:
def func(a, b=1, *args, d, e=True):
    print(args)
    
func(a=2, 5, 10 , 15, d=2)

SyntaxError: positional argument follows keyword argument (3384403498.py, line 4)

In [123]:
func(2, 5, 10 , 15, d=2)

TypeError: func() got an unexpected keyword argument 'd'

**Important**: If you want to use a specific keyword argument, you **must** use `*` or `*args` to exhaust the positional arguments. In the below, it may look like a keyword argument but it is not; we can call it without its keyword; it is infact a positional argument with a default value.

In [124]:
def  func(a, b, not_keyword_arg=10):
    print(a, b, not_keyword_arg)

func(1, 2, 3)

1 2 3


Whenever you see `*args` it's good to think of it as a thing that scoops up all other arguments.

In [125]:
def func1(a, b, *args, d):
    print(a, b, args, d)
    
func1(1, 2, 3, 4, 5)

TypeError: func1() missing 1 required keyword-only argument: 'd'

A good way to think about it is that python first assigns `a` and `b` to `1` and `2`. Then, it sees `*args` and it scoops up the remaining `3, 4` and `5`. So, it didn't find `d`. The only way we solve this is by specifying that `d=5`, i.e., to use a keyword argument.

Remember when we said that if you set a default parameter, then all parameters after that have to be specified? That is not actually true if we use `*args`.\
**Also, note that `e` below is *not* a positional argument. It is a keyword argument, but we don't need to specify a default value. We just need to specify a value when we call it.**

In [126]:
def func(a, b=20, *args, d=0, e): 
    print(a, b, args, d, e)
    
func(1, 2, 3, 4, 5, d=2, e=10) # we didnt have to specify arguments after 'b' because all remaining arguments will be shoved into 'args'

1 2 (3, 4, 5) 2 10


# 06 - Kwargs

We saw that you can scoop up all remaining positional arguments. Is there an equivalent counterpart for keyword arguments? Yes, it's called `**kwargs`, though that name is just convention. 
Here's some details:
- All remainings kwargs are scooped up into a dictionary.
- `**kwargs` can be specified even if the positional arguments have **not** been exhausted.
- No parameters can come after `**kwargs`.

Here's an example.

In [127]:
def func(*args, **kwargs):
    print(args)
    print(kwargs)

func(1, 2, 3, d=4, e=5, f=6)

(1, 2, 3)
{'d': 4, 'e': 5, 'f': 6}


If we want a named keyword argument as well as the scooper, then we will have to use the * stopper. Let's say we want `d` to be a named keyword argument:

In [128]:
def func(a, b, *, d, **kwargs):
    print(a, b, d, kwargs)
    
func(1, 2, x=10, y=20, d=5) # d must be specified, otherwise error.

1 2 5 {'x': 10, 'y': 20}


# 07 - Putting it all Together 

Let's take a look at python's `print()` function.

`print(*objects, sep='', end='\n', file=sys.stdout, flush=False)`

The `*objects` allows us to write many comma-separated statements to be printed. Since it is a`*` type parameter, it means that everything after must be keyword arguments. In this case, all of them have default values, therefore all of them are optional.

Typical Use Cases

Normally we use keyword-only arguments to modify the default behaviour of a function e.g.

In [129]:
def calc_hi_low_avg(*args, log_to_console=False):
    high = int(bool(args)) and max(args) # If no args passed (args tuple is empty), then its falsy and we return high as bool(False) which equals 0. 
                                         # If it's truthy, we immediately return high=max(args)  
    if log_to_console:
        print('this is the avg: .., the high.., and the low..')

# 08 - Simple Function Timer 

We want to create a simple function that can time how fast a function runs.

We want this function to be generic in the sense that it can be used to time any function (along with it's positional and keyword arguments), as well as specifying the number of the times the function should be timed, and the returns the average of the timings.

We'll call our function **time_it**, and it will need to have the following parameters:

* the function we want to time
* the positional arguments of the function we want to time (if any)
* the keyword-only arguments of the function we want to time (if any)
* the number of times we want to run this function

Note that in the line `fn(*args, **kwargs)`, we use `*` to unpack the args tuple and `**` to unpack the kwargs dictionary into the `fm`. If we didn't have them, then we will print out the literal tuple and dictionary, which is not what we want. 

**But note**: It did not change to a string - inside a function call, `*args` will unpack args into **separate positional arguments**. Normally when we've seen lines like `*b = 'longstring'`, we got back `b` as a `list`. But here, Python sees a `*` type and understands that it must be unpacked as function arguments, not as a list.

We can also see that `rep=5` is a keyword argument (because it follows `*args`) and it has a default value that can be overridden. We  just need to make sure that it `rep=2` is after all positional arguments i.e. after `1, 2, 3,`.

In [130]:
import time

def time_it(fn, *args, rep=5, **kwargs):
    start = time.perf_counter()
    for i in range(rep):
        fn(*args, **kwargs)
    end = time.perf_counter()
    return (end - start) / rep
        
time_it(print, 1, 2, 3, rep=2, sep='-')

1-2-3
1-2-3


3.7424500078486744e-05

More examples e.g. using *, mandatory keyword arguments etc. can be found in the full set of notes.

# 09 - Parameter Defaults - Beware

When we load a module, all the code is **executed**. This means that with lines like `a=10`, python creates an integer object with a memory address and then uses `a` as a reference. 

The same is true for functions. The function itself is not executed; rather, a function object is created in memory and referenced by a variable. If we create default parameters to that function, then those parameters are also created in memory and referenced. What this means is, when a user calls the function, the default value for the prior specified value is not re-evaluated at runtime. Instead, Python just goes looking for the reference. This means that we can call that function numerous times without having to worry about spending time to evaluate these function parameters.

**But beware:** If something like datetime.now() is passed as a default parameter within the module, then when the code is executed by a user, the datetime value will refer to when that module was created as opposed to when the user executed the code.

**So whats the typical solution pattern?**:

We still create a default parameter called `dt` but we set it to `None`. Then, within the function, we check if `dt=None`. If so, we can update it using `dt=datetime.now()`. When the user calls the function without specifying `dt`, the `if` statement will trigger and `datetime.now()` will execute during runtime. 

# 10 - Parameter Defaults - Beware Again

Another gotcha with parameter defaults comes with mutable types, and is an easy trap to fall into.

Again, you have to remember that function parameter defaults are evaluated once, when the function is defined (i.e. when the module is loaded, or in this Jupyter notebook, when we "execute" the function definition), and not every time the function is called.

Consider the following scenario.

We are creating a grocery list, and we want our list to contain consistently formatted data with name, quantity and measurement unit:

``
bananas (2 units)
grapes (1 bunch)
milk (1 liter)
python (1 medium-rare)
``

In [131]:
def add_item(name, quantity, unit, grocery_list):
    item_fmt = "{0} ({1} {2})".format(name, quantity, unit)
    grocery_list.append(item_fmt)
    return grocery_list

store_1 = []
store_2 = []

add_item('bananas', 2, 'units', store_1)
add_item('grapes', 1, 'bunch', store_1)
add_item('python', 1, 'medium-rare', store_2)

print(store_1)
print(store_2)

['bananas (2 units)', 'grapes (1 bunch)']
['python (1 medium-rare)']


Ok, working great. But let's make the function a little easier to use - if the user does not supply an existing grocery list to append the item to, let's just go ahead and default our `grocery_list` to an empty list hence starting a new shopping list:

In [132]:
def add_item(name, quantity, unit, grocery_list=[]):
    item_fmt = "{0} ({1} {2})".format(name, quantity, unit)
    grocery_list.append(item_fmt)
    return grocery_list

store_1 = add_item('bananas', 2, 'units')
add_item('grapes', 1, 'bunch', store_1)

print(store_1)

['bananas (2 units)', 'grapes (1 bunch)']


OK, so that seems to be working as expected.

Let's start our second list:

In [133]:
store_2 = add_item('milk', 1, 'gallon')
print(store_2)

['bananas (2 units)', 'grapes (1 bunch)', 'milk (1 gallon)']


??? What's going on? Our second list somehow contains the items that are in the first list.

What happened is that the returned value in the first call we made was the default grocery list - but remember that the list was created once and for all when the function was **created** not called. So everytime we call the function, that is the **same** list being used as the default (same memory address).

When we started out first list, we were adding item to that default list.

When we started our second list, we were adding items to the **same** default list (since it is the same object).

In [134]:
store_1 == store_2

True

We can avoid this problem using the same pattern as in the previous example we had with the default date time value. We use None as a default value instead, and generate a new empty list (hence starting a new list) if none was provided.

In [135]:
def add_item(name, quantity, unit, grocery_list=None):
    if not grocery_list:
        grocery_list = []
    item_fmt = "{0} ({1} {2})".format(name, quantity, unit)
    grocery_list.append(item_fmt)
    return grocery_list

store_1 = add_item('bananas', 2, 'units')
add_item('grapes', 1, 'bunch', store_1)

store_2 = add_item('milk', 1, 'gallon')

print(store_1)
print(store_2)

['bananas (2 units)', 'grapes (1 bunch)']
['milk (1 gallon)']


We can actually take advantage of this mutability. In the following example, we have a default argument called cache and thus is created when the function object is created and referenced to a memory address.

In [136]:
def factorial(n, cache={}):
    if n < 1:
        return 1
    elif n in cache:
        return cache[n]
    else:
        print('calculating {0}!'.format(n))
        result = n * factorial(n-1)
        cache[n] = result
        return result

Now, during runtime of this function call below, this cache is being updated with each recursion of the function call. The cache is now permanently updated somewhere in memory, containing a set of keys named 1, 2, 3, 4 and 5, with labels equivalent to their factorial. 

In [137]:
print(factorial(5))

calculating 5!
calculating 4!
calculating 3!
calculating 2!
calculating 1!
120


When we call this function again, we don't even use recursion. We go through the `elif` statement and just look for the key named `5`. It will have a value equal to 120.

In [138]:
print(factorial(5))

120


This pattern is called **memoisation**.

While this seems extremely elegant, it is not recommended since it requires more precaution as we're dealing with mutables. Also, there's a far more general and elegant way of doing this using decorators and closures.